### PR TITLE
DS 3558 Case-insensitive bot matching option

### PR DESCRIPTION
**Pull request by Frederic-Atmire**
Made case-insensitive matching for user-agents configurable to solve performance issues with large repositories. Also changed Pattern.CASE_INSENSITIVE to "toLowerCase()" because this gave an increase in performance.

Fix for jira issue: DS-3558

Also note that because of this PR, these configs are also now reloadable. (comment added at the suggestion of Tim Donohue)

### CODE REVIEW

**Review by bram-atmire — APPROVED** [review]
important performance fix. Looking for an additional non-Atmire person to review and approve this.

File: dspace/config/modules/usage-statistics.cfg [hunk 1]
```diff
@@ -35,3 +35,7 @@ usage-statistics.authorization.admin.workflow=true
 # (see query.filter.* for query filter options)
 # Default value is true.
 #usage-statistics.logBots = true
+
+# Enable/disable if a matching for a bot should be case sensitive
+# Setting this value to true will increase cpu usage, but bots will be found more accurately
+usage-statistics.bots.case-insensitive=true
```
Comment (tdonohue) [inline dspace/config/modules/usage-statistics.cfg]: Per the associated ticket, DS-3558, it's mentioned in the description that this setting should be "disabled by default". Has that been rethought?

**Comment by hardyoyo** [comment]
The testing file mentioned on DS-3558 would be helpful for testing this. I can confirm it compiles and deploys cleanly in Vagrant-DSpace. I'm trying to figure out a way to _test_ it...

**Review by tdonohue — COMMENTED** [review]
Added another minor comment here.  Per @hardyoyo's note, could we add a Unit test of the behavior to this PR perhaps? That'd give us more validation on the behavior (with the config on/off) at least.

File: dspace/config/modules/usage-statistics.cfg [hunk 2]
```diff
@@ -35,3 +35,7 @@ usage-statistics.authorization.admin.workflow=true
 # (see query.filter.* for query filter options)
 # Default value is true.
 #usage-statistics.logBots = true
+
+# Enable/disable if a matching for a bot should be case sensitive
+# Setting this value to true will increase cpu usage, but bots will be found more accurately
+usage-statistics.bots.case-insensitive=false
```
Comment (tdonohue) [inline dspace/config/modules/usage-statistics.cfg]: In a minor release (e.g. 6.1), we should comment out this setting by default. In minor releases, our configs should be backwards compatible, so if this config is missing, nothing should "break".

**Comment by mwoodiupui** [comment]
Curious:  why remove all the 'static's?

**Comment by hardyoyo** [comment]
I'm equally perplexed by the removal of the 'static's... trying to figure it out, can you help us?

**Comment by hardyoyo** [comment]
For what it's worth, this code compiles, and the tests all run. So... looking good, just trying to figure out the code now.

File: dspace-api/src/main/java/org/dspace/statistics/util/SpiderDetector.java [hunk 3]
```diff
@@ -271,6 +279,21 @@ public static boolean isSpider(String clientIP, String proxyIPs,
...
+        return spiderDetector.isSpiderAgent(clientIP, proxyIPs, hostname, agent);
+    }
+
+    public boolean isSpiderAgent(HttpServletRequest request)
+    {
+        return isSpiderAgent(request.getRemoteAddr(),
+                request.getHeader("X-Forwarded-For"),
+                request.getRemoteHost(),
+                request.getHeader("User-Agent"));
+    }
```
Comment (mwoodiupui) [inline dspace-api/src/main/java/org/dspace/statistics/util/SpiderDetector.java]: Is all this just to avoid mocking up an HttpServletRequest?

**Comment by hardyoyo** [comment]
I like performance improvements, and would love to get another one into 6.1, I'd approve this if there were a comment explaining the refactor in the commit that removes the statics. If the spider configs are now reloadable due to this PR, that's something that needs documenting.

**Comment by mwoodiupui** [comment]
There is a lot of change that I can't explain in terms of making matching case-insensitive or in terms of testing.  More documentation would help.

**Comment by hardyoyo** [comment]
@Frederic-Atmire, @bram-atmire I think it's clear from the comments above that we are *keenly* interested in seeing this patch get merged in time for DSpace 6.1. I think a few commit comments would be enough "documentation" here.

**Comment by Frederic-Atmire** [comment]
I removed the static's purely so I can indeed reload the configs, the only reason this was necessary was so I can write tests more easily, it was not intended to add any functionality. If needed I can rollback the static's, but then I have to remove the tests.

**Comment by tdonohue** [comment]
@Frederic-Atmire : thanks for the explanation. I'd rather see the PR as-is (with Unit Tests in here and the `static` statements removed) than doing a smaller PR with no Unit Tests.  I suspect others would agree, and simply wanted an explanation around the larger code changes (which now makes more sense, thanks).

**Comment by mwoodiupui** [comment]
It would be less intrusive to just mock the ConfigurationService and control what it shows to each test, but I'm not going to insist.

**Comment by hardyoyo** [comment]
So, @tdonohue, @mwoodiupui, are @Frederic-Atmire's comments above sufficient "documenation" for the changes in this PR?

**Comment by tdonohue** [comment]
@hardyoyo : I'm 👍 here, as it sounds like it has been tested and comes with Unit Tests. We should just update the description of this PR (or the ticket comments) to also note that because of this PR, these configs are also now reloadable.  (That'll just help if we ever wonder when that change occurred.)

**Comment by Frederic-Atmire** [comment]
To make the code a bit cleaner I now created a new class "SpiderDetectorServiceImpl", which will be used by the SpiderDetector (instead of using a reference to itself). With this change it's very easy to create a new "SpiderDetectorService" since this is a Spring bean (which opens up the option to make a custom one). I hope this doesn't make things too complicated.

**Review by mwoodiupui — APPROVED** [review]

**Pull request merged**